src/Common/src/System/Runtime/InteropServices/StringBuffer.cs

```diff
@@ -0,0 +1,319 @@
+    /// comparisons.
+    /// 
+    /// A more performant replacement for StringBuilder when performing native interop.
+    /// </summary>
+    /// <remarks>
+    /// Suggested use through P/Invoke: define DllImport arguments that take a character buffer as IntPtr.
+    /// NativeStringBuffer has an implicit conversion to IntPtr.
+    /// </remarks>
+    internal class StringBuffer : NativeBuffer
+    {
+        private ulong _length;
+
+        /// <summary>
+        /// Instantiate the buffer with capacity for at least the specified number of characters. Capacity
+        /// includes the trailing null character.
+        /// </summary>
+        public StringBuffer(ulong initialCapacity = 0)
+            : base(initialCapacity)
+        {
+        }
+
+        /// <summary>
+        /// Instantiate the buffer with a copy of the specified string.
+        /// </summary>
+        public unsafe StringBuffer(string initialContents)
+            : base(0)
+        {
+            // We don't pass the count of bytes to the base constructor, appending will
+            // initialize to the correct size for the specified initial contents.
+            if (initialContents != null)
+            {
+                Append(initialContents);
+            }
+        }
+
+        /// <summary>
+        /// Get/set the character at the given index.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if attempting to index outside of the buffer length.</exception>
+        public new unsafe char this[ulong index]
+        {
+            get
+            {
+                if (index >= _length) throw new ArgumentOutOfRangeException("index");
+                return CharPointer[index];
+            }
+            set
+            {
+                if (index >= _length) throw new ArgumentOutOfRangeException("index");
+                CharPointer[index] = value;
+            }
+        }
+
+        /// <summary>
+        /// Character capacity of the buffer. Includes the count for the trailing null character.
+        /// </summary>
+        public ulong CharCapacity
+        {
+            get
+            {
+                ulong byteCapacity = ByteCapacity;
+                return byteCapacity == 0 ? 0 : byteCapacity / sizeof(char);
+            }
+        }
+
+        /// <summary>
+        /// Ensure capacity in characters is at least the given minimum.
+        /// </summary>
+        /// <exception cref="OutOfMemoryException">Thrown if unable to allocate memory when setting.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if attempting to set <paramref name="Capacity"/> to a value that is larger than the maximum addressable memory.</exception>
+        public void EnsureCharCapacity(ulong minCapacity)
+        {
+            if (minCapacity > (ulong.MaxValue / sizeof(char))) throw new ArgumentOutOfRangeException("minCapacity");
```

Comment:
Not sure this check is really valuable.  The only time it'll throw is if the caller is requesting more than 9 exabytes of capacity, in which case the next line is going to OOM.